File: core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/ChannelMarkerManager.scala
package edu.uci.ics.amber.engine.architecture.worker

import edu.uci.ics.amber.engine.architecture.messaginglayer.InputGateway
import edu.uci.ics.amber.engine.architecture.worker.ChannelMarkerManager.MarkerContext
import edu.uci.ics.amber.engine.common.{AmberLogging, CheckpointState}
import edu.uci.ics.amber.engine.common.ambermessage.{
  ChannelMarkerPayload,
  NoAlignment,
  RequireAlignment
}
import edu.uci.ics.amber.engine.common.virtualidentity.{
  ActorVirtualIdentity,
  ChannelIdentity,
  ChannelMarkerIdentity
}

import scala.collection.mutable

object ChannelMarkerManager {
  final case class MarkerContext(marker: ChannelMarkerPayload, fromChannel: ChannelID)
}

class ChannelMarkerManager(inputGateway: InputGateway, val actorId: ActorVirtualIdentity, inputGateway: InputGateway)
    extends AmberLogging {

  private val markerReceived =
    new mutable.HashMap[ChannelMarkerIdentity, Set[ChannelIdentity]]().withDefaultValue(Set())

  private var markerContext: MarkerContext = _

  val checkpoints = new mutable.HashMap[ChannelMarkerIdentity, CheckpointState]()

  def setContext(marker: ChannelMarkerPayload, from: ChannelID): Unit = {
    markerContext = MarkerContext(marker, from)
  }

  def getContext: MarkerContext = markerContext

  /**
    * Determines if an epoch marker is fully received from all relevant senders within its scope.
    * This method checks if the epoch marker, based on its type, has been received from all necessary channels.
    * For markers requiring alignment, it verifies receipt from all senders in the scope. For non-aligned markers,
    * it checks if it's the first received marker. Post verification, it cleans up the markers.
    *
    * @return true if the epoch marker is fully received from all senders in the scope, false otherwise.
    */
  def isMarkerAligned: Boolean = {
    assert(markerContext != null)
    val markerId = markerContext.marker.id
    markerReceived.update(markerId, markerReceived(markerId) + markerContext.fromChannel)
    // check if the epoch marker is completed
    val markerReceivedFromAllChannels =
      getChannelsWithinScope.subsetOf(markerReceived(markerId))
    val epochMarkerCompleted = markerContext.marker.markerType match {
      case RequireAlignment => markerReceivedFromAllChannels
      case NoAlignment      => markerReceived(markerId).size == 1 // only the first marker triggers
    }
    if (markerReceivedFromAllChannels) {
      markerReceived.remove(markerId) // clean up if all markers are received
    }
    epochMarkerCompleted
  }

  private def getChannelsWithinScope: Set[ChannelIdentity] = {
    assert(markerContext != null)
    val upstreams = markerContext.marker.scope.filter(_.toWorkerId == actorId)
    inputGateway.getAllChannels
      .map(_.channelId)
      .filter { id =>
        upstreams.contains(id)
      }
      .toSet
  }

}

File: core/amber/src/main/python/core/architecture/managers/channel_marker_manager.py
from collections import defaultdict
from typing import Set, Dict

from proto.edu.uci.ics.amber.core import ActorVirtualIdentity, ChannelIdentity
from proto.edu.uci.ics.amber.engine.architecture.rpc import (
    ChannelMarkerPayload,
    ChannelMarkerType,
)


class ChannelMarkerManager:
    def __init__(self, actor_id: ActorVirtualIdentity, input_gateway):
        self.actor_id = actor_id
        self.input_gateway = input_gateway
        self.marker_received: Dict[str, Set[ChannelIdentity]] = defaultdict(set)

    def is_marker_aligned(
        self, from_channel: ChannelIdentity, marker: ChannelMarkerPayload
    ):
        marker_id = marker.id
        self.marker_received[marker_id].add(from_channel)

        marker_received_from_all_channels = self.get_channels_within_scope(
            marker
        ).issubset(self.marker_received[marker_id])

        if marker.marker_type == ChannelMarkerType.REQUIRE_ALIGNMENT:
            epoch_marker_completed = marker_received_from_all_channels
        elif marker.marker_type == ChannelMarkerType.NO_ALIGNMENT:
            epoch_marker_completed = (
                len(self.marker_received[marker_id]) == 1
            )  # Only the first marker triggers
        else:
            raise ValueError(f"Unsupported marker type: {marker.marker_type}")

        if marker_received_from_all_channels:
            del self.marker_received[marker_id]  # Clean up if all markers are received

        return epoch_marker_completed

    def get_channels_within_scope(self, marker: ChannelMarkerPayload):
        upstreams = {
            channel_id
            for channel_id in marker.scope
            if channel_id.to_worker_id == self.actor_id
        }
        return {
            channel_id
            for channel_id in self.input_gateway.get_all_channel_ids()
            if channel_id in upstreams
        }
